src/NHibernate/Dialect/FirebirdDialect.cs

```diff
@@ -417,7 +417,7 @@ private void OverrideStandardHQLFunctions()
 			RegisterFunction("nullif", new StandardSafeSQLFunction("nullif", 2));
 			RegisterFunction("lower", new StandardSafeSQLFunction("lower", NHibernateUtil.String, 1));
 			RegisterFunction("upper", new StandardSafeSQLFunction("upper", NHibernateUtil.String, 1));
-			RegisterFunction("mod", new StandardSafeSQLFunction("mod", NHibernateUtil.Double, 2));
+			RegisterFunction("mod", new StandardSafeSQLFunction("mod", NHibernateUtil.Int32, 2));
```

Comment:
Ideally depending on the dialect it should be either Int32 (FB <3.0) or Int64 (FB >3.0).

Reply:
From [the documentation](https://firebirdsql.org/refdocs/langrefupd25-intfunc-mod.html) the result varies depending on the arguments. Also the arguments are rounded before appling the mod operator, which is not consistent with .NET and other databases. In the current master, the mod operator is never executed on the DB, which allows to consistently calculate the mod operation on the client side, but it prevents us from using it in a subquery. In this PR the mod operator is currently always executed on the database, which will not work correctly when using decimal values as arguments as all dialects have `Int32` set as the return type. In order to fix that I will make custom `ISQLFunction` classes for each dialect, which will have the information whether the mod operator can be translated for the given arguments or not. For example, in firebird when one of the arguments is decimal the mod operator will not be translated to sql in order to keep the consistency. As mod is probably not the only function which its result type varies depending on the arguments I am setting again to WIP, until the mod and other similar functions are fixed.

Reply:
Have you tried supplying `null` as the type? In such case, NHibernate infers the type from the first argument of the function. But maybe you need to do that according to the second argument, or considering both, in which case some changes a bit deeper will be required. (See [`FindFunctionReturnType`](https://github.com/nhibernate/nhibernate-core/blob/94a94da5a3f8b5436f1b6604995c8a5acd2e63aa/src/NHibernate/Hql/Ast/ANTLR/SessionFactoryHelperExtensions.cs#L70).)

Reply:
Yes I tried, but in case of `mod` function the returning type depends on both parameters. I am intending to add an overload for `FindFunctionReturnType` which will take all function arguments and also extend `ISQLFunction` with a transitional interface that will do the same for `ReturnType` mehod. Unfurtunately, the new `ReturnType` method will not be enough, because the same method is called also for a hql query that may not have the information about the type. For example:
```
var q = session.CreateQuery("select mod(a.Id, :p1) from Animal a")
               .SetParameter("p1", 2.1m);
```
In the above query, the hql string is analyzed (`FindFunctionReturnType` gets called) when the `CreateQuery` method is called, which means that we do not have the type of the paramater. For this reason the transitional `ISQLFunction` interface will need to have an additional property `DefaultReturnType` which will be used when arguments types cannot be determined. So for the given example, all values will be converted to `int`, because the default return type for the `mod` function is set to `Int32`, which is also how it currently works. For a linq query the parameter will be wrapped with a `transparentcast` in order to preserve the type when building the hql tree, so that `FindFunctionReturnType` will be able to determine all parameter types.